man/ttultest.1
.TH "TTULTEST" 1 "2010-08-05" "Man Page" "Tokyo Tyrant"

.SH NAME
ttultest \- test cases of the update log API

.SH DESCRIPTION
.PP
hey.

.SH SEE ALSO
.PP
.BR ttserver (1),
.BR ttulmgr (1),
.BR ttutil (3),
.BR tculog (3),
.BR tcrdb (3)


ChangeLog
2010-02-02  FAL Labs  <info@fallabs.com>

	- site documents were modified.

	- Release: 1.1.41

2010-01-01  FAL Labs  <info@fallabs.com>

	* ttserver.c (proc): a bug related to initialization of the skeleton was fixed.

	* tculog.c (tculogadbaddint, tculogadbadddouble): efficiency of zero was improved.

	* tcrmgr.c (runput, procput): addint and adddouble were added.

	- Release: 1.1.40

2009-12-07  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttservstart): error handler for the accept call was added.

	* ttserver.c (proc): the system connection limit is now reset.

	- Release: 1.1.39

2009-11-05  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_mc_append, do_mc_prepend): new functions.

	- Release: 1.1.38

2009-10-20  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_stat): "fatal" attribute was added.

	* ttutil.c (ttwaitsock): new function.

	* ttutil.c (ttopensock, ttsocksend, ttsockgetc): "errno" is now copied just in case.

	* tcrdb.c (tcrdbparasearchworker): a bug of race condition was fixed.

	- Release: 1.1.37

2009-10-06  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_strstr): new function.

	* ttserver.c (main, proc): "-mul" option was added.

	- Release: 1.1.36

2009-09-18  FAL Labs  <info@fallabs.com>

	* tculog.c (tculogadbputshl, tculogadbputshlproc): new functions.

	* scrext.c (iterrec): a bug related to handling the return value was fixed.

	* ttserver.c (do_http_post): "tcadbmisc" is now supported.

	* ttserver.c (do_stat, do_mc_stats): counts of command execution are now output.

	* tcrdb.c (tcrdbparasearch, tcrdbparasearchworker): new functions.

	- Release: 1.1.35

2009-08-04  FAL Labs  <info@fallabs.com>

	* tcrdb.c (tcrdbput, tcrdbout, tcrdbget): reconnect mechanism was enhanced.

	* tcrdb.c (tcrdbreconnect): delay mecahism was added.

	* tculog.c (tculrdnew): a bug related to timestamp detection was fixed.

	* tcrmgr.c (runimporttsv, procimporttsv): "-sep" option was added.

	* ttserver.c (do_log): a bug of buffer overrun was fixed.

	- Release: 1.1.34

2009-07-18  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_bit): new function.

	* tcrdb.c (tcrdbqryhint, tcrdbmetasearch): new functions.

	- Release: 1.1.33

2009-07-15  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_mget, do_fwmkeys, do_misc): alignment violations were cleared.

	* tcrmttest.c (runtypical, proctypical, threadtypical): new functions.

	- Release: 1.1.32

2009-07-09  FAL Labs  <info@fallabs.com>

	* tculog.c (tculogwrite, tculrdread): the master ID is now recorded.

	* ttserver.c (main): the default value of the "-ulim" option was modified.

	* ttserver.c (proc): update logging and replicaion are now forbidden without the SID.

	* ttserver.c (do_repl): the NOP command is now throwed while circular replication.

	- Release: 1.1.31

2009-07-05  FAL Labs  <info@fallabs.com>

	* scrext.c (scrextkill): new function.

	* scrext.c (serv_lock, serv_unlock): lock mechanism was modified with TCMDB.

	- Release: 1.1.30

2009-05-28  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_http_options): new function.

	* tcrtest.c (proctable): a test of full-text search was added.

	- Release: 1.1.29

2009-05-20  FAL Labs  <info@fallabs.com>

	* ttserver.c (do_task): a potential bug of memory leak was fixed.

	* tcrdb.c (tcrdbnew): a bug related to initialization was fixed.

	* scrext.c (serv_fwmkeys, serv_regex): new functions.

	- Release: 1.1.28

2009-05-11  FAL Labs  <info@fallabs.com>

	* tcrtest.c (runtable, proctable): "-exp" option was added.

	* ttskelnull.c: new file.

	* tokyotyrant.idl: new file.

	- Release: 1.1.27

2009-05-09  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttservsettermhandler): new function.

	* ttserver.c (do_term): new function.

	- Release: 1.1.26

2009-05-06  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttcmdidtostr): new function.

	* tculog.c (tcreplopen): a bug of data size of the server ID was fixed.

	* tcrdb.c (tcrdbsetmst, tcrdbsetmst2): the timestamp argument was added.

	* tcrmgr.c (procrepl): command names are now printed.

	* ttulmgr.c (procexport): command names are now printed.

	- Release: 1.1.25

2009-04-25  FAL Labs  <info@fallabs.com>

	* ttutil.c (ttbreakservexpr): new function.

	* tcrdb.c (tcrdbopen2, tcrdbsetmst2, tcrdbtune): new functions.

	* tcrdb.c (tcrdbsetecode, tcrdblockmethod, tcrdbunlockmethod): new functions.

	* ttserver.c (proc): dynamic linking for the skeleton database was added.

	* ttskelmock.c, ttskelproxy.c: new files.

	- Release: 1.1.24

2009-04-07  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_eval, serv_split): new functions.

	* ttserver.c (do_repl): efficiency on dual-master replication was improved.

	* ttserver.c (do_optimize): new function.

	* ttserver.c (do_log): message format was modified.

	* tcrdb.c (tcrdboptimize): new function.

	* tcrdb.c (tcrdbvanish, tcrdbcopy, tcrdbrestore): each magic number was drifted.

	* tcrdb.c (tcrdbrestore, tcrdbsetmst): an argument for options was added.

	* tculog.c (tculogadbredo): the flag of consistency checking was separated.

	- Release: 1.1.23

2009-04-06  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_mapreduce): the parameter for the target keys became an option.

	- Release: 1.1.22

2009-04-06  FAL Labs  <info@fallabs.com>

	* scrext.c (serv_mapreduce): the parameter for the session ID was abolished.

	- Release: 1.1.21

2009-04-01  FAL Labs  <info@fallabs.com>

	* configure.in: a bug related to support for traditional shells was fixed.

	* scrext.c (serv_stashputkeep, serv_stashputcat): new functions.

	* scrext.c (serv_mapreduce, serv_mapreducemapemit): new functions.

	* ttserver.c (do_put, do_out, do_get): parameter validation was enhanced.

	- Release: 1.1.20

2009-03-14  FAL Labs  <info@fallabs.com>

	* tcrdb.c (tcrdbqrysearchget): the option for no update log was added.

	* tcrdb.c (tcrdbqrysearchcount): new function.

	* tculog.c (tcreplread): timeout mechanism was added.

	* ttserver.c (do_repl): performance was improved.

	- Release: 1.1.19

2009-03-14  FAL Labs  <info@fallabs.com>


ttultest.c
/*************************************************************************************************
 * The test cases of the update log API
 *                                                               Copyright (C) 2006-2010 FAL Labs
 * This file is part of Tokyo Tyrant.
 * Tokyo Tyrant is free software; you can redistribute it and/or modify it under the terms of
 * the GNU Lesser General Public License as published by the Free Software Foundation; either
 * version 2.1 of the License or any later version.  Tokyo Tyrant is distributed in the hope
 * that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 * You should have received a copy of the GNU Lesser General Public License along with Tokyo
 * Tyrant; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA.
 *************************************************************************************************/


#include <tculog.h>
#include "myconf.h"

#define RECBUFSIZ      32                // buffer for records

typedef struct {                         // type of structure for read thread
  TCULRD *ulrd;
  int id;
  int rnum;
} TARGREAD;


/* global variables */
const char *g_progname;                  // program name


/* function prototypes */
int main(int argc, char **argv);
static void usage(void);
static void iprintf(const char *format, ...);
static void eprint(TCULOG *ulog, const char *func);
static int runwrite(int argc, char **argv);
static int runread(int argc, char **argv);
static int runthread(int argc, char **argv);
static int procwrite(const char *base, int rnum, int64_t limsiz, bool as);
static int procread(const char *base, uint64_t ts, bool pm);
static int procthread(const char *base, int tnum, int rnum, int64_t limsiz, bool as);


/* main routine */
int main(int argc, char **argv){
  g_progname = argv[0];
  srand((unsigned int)(tctime() * 1000) % UINT_MAX);
  if(argc < 2) usage();
  int rv = 0;
  if(!strcmp(argv[1], "write")){
    rv = runwrite(argc, argv);
  } else if(!strcmp(argv[1], "read")){
    rv = runread(argc, argv);
  } else if(!strcmp(argv[1], "thread")){
    rv = runthread(argc, argv);
  } else {
    usage();
  }
  return rv;
}


/* print the usage and exit */
static void usage(void){
  fprintf(stderr, "%s: test cases of the remote database API of Tokyo Tyrant\n", g_progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "usage:\n");
  fprintf(stderr, "  %s write [-lim num] [-as] base rnum\n", g_progname);
  fprintf(stderr, "  %s read [-ts num] [-pm] base\n", g_progname);
  fprintf(stderr, "  %s thread [-lim num] [-as] base tnum rnum\n", g_progname);
  fprintf(stderr, "\n");
  exit(1);
}


/* print formatted information string and flush the buffer */
static void iprintf(const char *format, ...){
  va_list ap;
  va_start(ap, format);
  vprintf(format, ap);
  fflush(stdout);
  va_end(ap);
}


/* print error message of abstract database */
static void eprint(TCULOG *ulog, const char *func){
  fprintf(stderr, "%s: %s: error:\n", g_progname, func);
}


/* parse arguments of write command */
static int runwrite(int argc, char **argv){
  char *base = NULL;
  char *rstr = NULL;
  int64_t limsiz = 0;
  bool as = false;
  for(int i = 2; i < argc; i++){
    if(!base && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-lim")){
        if(++i >= argc) usage();
        limsiz = tcatoi(argv[i]);
      } else if(!strcmp(argv[i], "-as")){
        as = true;
      } else {
        usage();
      }
    } else if(!base){
      base = argv[i];
    } else if(!rstr){
      rstr = argv[i];
    } else {
      usage();
    }
  }
  if(!base || !rstr) usage();
  int rnum = tcatoi(rstr);
  if(rnum < 1) usage();
  int rv = procwrite(base, rnum, limsiz, as);
  return rv;
}


/* parse arguments of read command */
static int runread(int argc, char **argv){
  char *base = NULL;
  uint64_t ts = 0;
  bool pm = false;
  for(int i = 2; i < argc; i++){
    if(!base && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-ts")){
        if(++i >= argc) usage();
        ts = ttstrtots(argv[i]);
      } else if(!strcmp(argv[i], "-pm")){
        pm = true;
      } else {
        usage();
      }
    } else if(!base){
      base = argv[i];
    } else {
      usage();
    }
  }
  if(!base) usage();
  int rv = procread(base, ts, pm);
  return rv;
}


/* parse arguments of thread command */
static int runthread(int argc, char **argv){
  char *base = NULL;
  char *tstr = NULL;
  char *rstr = NULL;
  int64_t limsiz = 0;
  bool as = false;
  for(int i = 2; i < argc; i++){
    if(!base && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-lim")){
        if(++i >= argc) usage();
        limsiz = tcatoi(argv[i]);
      } else if(!strcmp(argv[i], "-as")){
        as = true;
      } else {
        usage();
      }
    } else if(!base){
      base = argv[i];
    } else if(!tstr){
      tstr = argv[i];
    } else if(!rstr){
      rstr = argv[i];
    } else {
      usage();
    }
  }
  if(!base || !tstr || !rstr) usage();
  int tnum = tcatoi(tstr);
  int rnum = tcatoi(rstr);
  if(tnum < 1 || rnum < 1) usage();
  int rv = procthread(base, tnum, rnum, limsiz, as);
  return rv;
}


/* perform write command */
static int procwrite(const char *base, int rnum, int64_t limsiz, bool as){
  iprintf("<Writing Test>\n  base=%s  rnum=%d  limsiz=%lld  as=%d\n\n",
          base, rnum, (long long)limsiz, as);
  bool err = false;
  double stime = tctime();
  TCULOG *ulog = tculognew();
  if(as && !tculogsetaio(ulog)){
    eprint(ulog, "tculogsetaio");
    err = true;
  }
  if(!tculogopen(ulog, base, limsiz)){
    eprint(ulog, "tculogopen");
    err = true;
  }
  int sid = getpid() & UINT16_MAX;
  for(int i = 1; i <= rnum; i++){
    char buf[RECBUFSIZ];
    int len = sprintf(buf, "%08d", i);
    if(!tculogwrite(ulog, 0, sid, sid, buf, len)){
      eprint(ulog, "tculogwrite");
      err = true;
      break;
    }
    if(rnum > 250 && i % (rnum / 250) == 0){
      putchar('.');
      fflush(stdout);
      if(i == rnum || i % (rnum / 10) == 0) iprintf(" (%08d)\n", i);
    }
  }
  if(!tculogclose(ulog)){
    eprint(ulog, "tculogclose");
    err = true;
  }
  tculogdel(ulog);
  iprintf("time: %.3f\n", tctime() - stime);
  iprintf("%s\n\n", err ? "error" : "ok");
  return err ? 1 : 0;
}


/* perform read command */
static int procread(const char *base, uint64_t ts, bool pm){
  if(!pm)
    iprintf("<Reading Test>\n  base=%s  ts=%llu  pm=%d\n\n", base, (unsigned long long)ts, pm);
  bool err = false;
  double stime = tctime();
  TCULOG *ulog = tculognew();
  if(!tculogopen(ulog, base, 0)){
    eprint(ulog, "tculogopen");
    err = true;
  }
  TCULRD *ulrd = tculrdnew(ulog, ts);
  if(ulrd){
    const char *rbuf;
    int rsiz;
    uint64_t rts;
    uint32_t rsid, rmid;
    int i = 1;
    while((rbuf = tculrdread(ulrd, &rsiz, &rts, &rsid, &rmid)) != NULL){
      if(pm){
        printf("%llu\t%u:%u\t", (unsigned long long)rts, (unsigned int)rsid, (unsigned int)rmid);
        for(int i = 0; i < rsiz; i++){
          if(i > 0) putchar(' ');
          printf("%02X", ((unsigned char *)rbuf)[i]);
        }
        putchar('\n');
      } else {
        if(i % 1000 == 0){
          putchar('.');
          fflush(stdout);
          if(i % 25000 == 0) iprintf(" (%08d)\n", i);
        }
      }
      i++;
    }
    if(!pm) iprintf(" (%08d)\n", i - 1);
    tculrddel(ulrd);
  }
  if(!tculogclose(ulog)){
    eprint(ulog, "tculogclose");
    err = true;
  }
  tculogdel(ulog);
  if(!pm){
    iprintf("time: %.3f\n", tctime() - stime);
    iprintf("%s\n\n", err ? "error" : "ok");
  }
  return err ? 1 : 0;
}


/* thread the read function */
static void *threadread(void *targ){
  TCULRD *ulrd = ((TARGREAD *)targ)->ulrd;
  int id = ((TARGREAD *)targ)->id;
  int rnum = ((TARGREAD *)targ)->rnum;
  const char *rbuf;
  int rsiz;
  uint64_t rts;
  uint32_t rsid, rmid;
  int i = 1;
  while(i <= rnum){
    while((rbuf = tculrdread(ulrd, &rsiz, &rts, &rsid, &rmid)) != NULL){
      if(id == 0 && rnum > 250 && i % (rnum / 250) == 0){
        putchar('.');
        fflush(stdout);
        if(i == rnum || i % (rnum / 10) == 0) iprintf(" (%08d)\n", i);
      }
      i++;
    }
    tcsleep(0.01);
  }
  return NULL;
}


/* perform thread command */
static int procthread(const char *base, int tnum, int rnum, int64_t limsiz, bool as){
  iprintf("<Threading Test>\n  base=%s  tnum=%d  rnum=%d  limsiz=%lld  as=%d\n\n",
          base, tnum, rnum, (long long)limsiz, as);
  bool err = false;
  double stime = tctime();
  TCULOG *ulog = tculognew();
  if(as && !tculogsetaio(ulog)){
    eprint(ulog, "tculogsetaio");
    err = true;
  }
  if(!tculogopen(ulog, base, limsiz)){
    eprint(ulog, "tculogopen");
    err = true;
  }
  TARGREAD targs[tnum];
  pthread_t threads[tnum];
  for(int i = 0; i < tnum; i++){
    targs[i].ulrd = tculrdnew(ulog, 0);
    targs[i].id = i;
    targs[i].rnum = rnum;
    if(!targs[i].ulrd){
      eprint(ulog, "tculrdnew");
      targs[i].id = -1;
      err = true;
    } else if(pthread_create(threads + i, NULL, threadread, targs + i) != 0){
      eprint(ulog, "pthread_create");
      targs[i].id = -1;
      err = true;
    }
  }
  int sid = getpid() & UINT16_MAX;
  for(int i = 1; i <= rnum; i++){
    char buf[RECBUFSIZ];
    int len = sprintf(buf, "%08d", i);
    if(!tculogwrite(ulog, 0, sid, sid, buf, len)){
      eprint(ulog, "tculogwrite");
      err = true;
      break;
    }
    if(rnum > 250 && i % (rnum / 10) == 0) tcsleep(0.1);
  }
  for(int i = 0; i < tnum; i++){
    if(targs[i].id == -1) continue;
    void *rv;
    if(pthread_join(threads[i], &rv) != 0){
      eprint(ulog, "pthread_join");
      err = true;
    } else if(rv){
      err = true;
    }
    tculrddel(targs[i].ulrd);
  }
  if(!tculogclose(ulog)){
    eprint(ulog, "tculogclose");
    err = true;
  }
  tculogdel(ulog);
  iprintf("time: %.3f\n", tctime() - stime);
  iprintf("%s\n\n", err ? "error" : "ok");
  return err ? 1 : 0;
}



// END OF FILE
